### PR TITLE
http3: add quic hints

### PR DESCRIPTION
**Pull request by alyssawilk**
Allowing auto_http to assume some domains support HTTP/3 without trying TCP first.

Risk Level: low
Testing: unit test
Docs Changes: API docs
Release Notes: not yet
Fixes https://github.com/envoyproxy/envoy/issues/19427

### CODE REVIEW

File: api/envoy/config/core/v3/protocol.proto
Comment (RyanTheOptimist) [inline api/envoy/config/core/v3/protocol.proto:134]: Presumably if we supported both "h3" and "h3-2" we would add some sort of alpn field to this message?

File: source/common/http/alternate_protocols_cache_manager_impl.cc
Comment (RyanTheOptimist) [inline source/common/http/alternate_protocols_cache_manager_impl.cc:62]: nit: std::chrono::day(7)?

File: source/common/http/alternate_protocols_cache_manager_impl.cc
Comment (RyanTheOptimist) [inline source/common/http/alternate_protocols_cache_manager_impl.cc:65]: Should we do any sort of validation that we don't already have an entry for |origin|? (or just log in this case?)

**Review by RyanTheOptimist — DISMISSED** [review]
Nice!

File: api/envoy/config/core/v3/protocol.proto
Comment (alyssawilk) [reply api/envoy/config/core/v3/protocol.proto:134]: Yeah.  Ditto alternate ports and hostnames.  When/if we add it we can just default to h3

File: source/common/http/alternate_protocols_cache_manager_impl.cc
Comment (alyssawilk) [reply source/common/http/alternate_protocols_cache_manager_impl.cc:62]: yeah that'd be more clear =P

File: source/common/http/alternate_protocols_cache_manager_impl.cc
Comment (alyssawilk) [reply source/common/http/alternate_protocols_cache_manager_impl.cc:65]: we can have an entry - as documented, if it's in cache we'll preserve the cache entry over the hard-coded entry.  I didn't think that was worth logging since it should be the common case and would be kinda spammy

File: api/envoy/config/core/v3/protocol.proto
Comment (snowp) [inline api/envoy/config/core/v3/protocol.proto:173]: I was expecting to see this bit of documentation above while I was reading it, maybe it should be in both places?

File: source/common/http/alternate_protocols_cache_manager_impl.cc
Comment (snowp) [inline source/common/http/alternate_protocols_cache_manager_impl.cc:62]: Is 7 days a standard value here? Maybe we should explicitly name this in the docs? It's implict in the `ma` in the example, but might not be obvious

This also seems to mean that if someone pushed config to pre-fill this, then reverted the config, we'd retain a cache entry for that for 7 days. This is probably fine but maybe worth calling out so nobody gets confused about quic being used for a connection that isn't prefilled with the current config?

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this makes sense to me, a few suggestions on things to cover in docs

File: source/common/http/alternate_protocols_cache_manager_impl.cc
Comment (RyanTheOptimist) [inline source/common/http/alternate_protocols_cache_manager_impl.cc:62]: nit: this looks mis-indented, but I guess check-format is good with it? Do you understand why?

**Review by RyanTheOptimist — APPROVED** [review]
LGTM

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Comment by alyssawilk** [comment]
cc @markdroth ping?

File: api/envoy/config/core/v3/protocol.proto
Comment (markdroth) [inline api/envoy/config/core/v3/protocol.proto:135]: Should this be nested inside of `AlternateProtocolsCacheOptions`?  Or is there a reason to think this might be used somewhere else in the future?

My concern here is that we seem to have a bunch of messages that are HTTP-specific but do not have "http" in the name anywhere, which seems like a bit of namespace pollution.  I would have chosen a different name for `AlternateProtocolsCacheOptions` too, but at least that message already exists here, so nesting this new message inside of it avoids making the problem worse.

Comment (alyssawilk) [reply api/envoy/config/core/v3/protocol.proto:135]: I can move it over if we prefer nesting.  This is unlikely to be used elsewhere but I've seen nesting be problematic often enough I generally lean against.  Your call!

Comment (markdroth) [reply api/envoy/config/core/v3/protocol.proto:135]: I'd prefer to nest it if we don't think it's going to be used elsewhere.  Thanks!

**Review by markdroth — COMMENTED** [review]
/lgtm api

**Comment by alyssawilk** [comment]
@snowp can I get a maintainer stamp on the API changes?

**Pull request merged**
